webapp/sources/rudder/rudder-web/src/main/scala/com/normation/rudder/web/components/RuleGrid.scala

```diff
@@ -566,10 +566,11 @@ class RuleGrid(
     val t0 = System.currentTimeMillis
 
     val nodes = groupsLib.getNodeIds(line.rule.targets, nodesInfo)
+    val nodeMods = nodesInfo.values.map(node => if(nodes.contains(node.id)) node.policyMode else None).toSet
```

Comment:
i think this is the opposite of what should be done: nodesInfos is a map, so getting the NodeInfo from the nodeid is basically O(1), getting all nodeinfo is linear in the size of nodes target
here, you take all the nodeinfos, to take their nodeis, and traverse once per node to get the policy mode, so complexity is nodes target * total node /2

Reply:
nodesInfo.collect { case (nodeId, nodeinfo) if nodes.contains(nodeId) =>  nodeinfo.policyMode }

Reply:
or something like that